### ISSUE TITLE
Amazon EC2 instance crashes on numeric interface ids

### ISSUE THREAD

**Issue by koleini** [mirage/mirage-net-xen#22]
```
Xen Minimal OS!
  start_info: 0x1110000(VA)
    nr_pages: 0xf0000
  shared_inf: 0x7de83000(MA)
     pt_base: 0x1113000(VA)
nr_pt_frames: 0xd
    mfn_list: 0x990000(VA)
   mod_start: 0x0(VA)
     mod_len: 0
       flags: 0x0
    cmd_line: root=/dev/sda1 ro 4
  stack:      0x94f860-0x96f860
MM: Init
      _text: 0x0(VA)
     _etext: 0x5ffbd(VA)
   _erodata: 0x78000(VA)
     _edata: 0x80ae0(VA)
stack start: 0x94f860(VA)
       _end: 0x98fe68(VA)
  start_pfn: 1123
    max_pfn: f0000
Mapping memory range 0x1400000 - 0xf0000000
setting 0x0-0x78000 readonly
skipped 0x1000
MM: Initialise page allocator for 189c000(189c000)-f0000000(f0000000)
MM: done
Demand map pfns at f0001000-20f0001000.
Heap resides at 20f0002000-40f0002000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 0xf0001000.
Initialising scheduler
Thread "Idle": pointer: 0x20f0002010, stack: 0x2050000
Initialising xenbus
Thread "xenstore": pointer: 0x20f00027c0, stack: 0x2060000
Dummy main: start_info=0x96f960
Thread "main": pointer: 0x20f0002f70, stack: 0x2070000
"main" "root=/dev/sda1" "ro" "4" 
vbd 2049 is hd0
******************* BLKFRONT for device/vbd/2049 **********


backend at /local/domain/0/backend/vbd/2255/2049
26624 sectors of 512 bytes
**************************
vbd 2064 is hd1
******************* BLKFRONT for device/vbd/2064 **********


backend at /local/domain/0/backend/vbd/2255/2064
8377344 sectors of 512 bytes
**************************
[H[J

    GNU GRUB  version 0.97  (3932160K lower / 0K upper memory)



[m[4;2H+-------------------------------------------------------------------------+[5;2H|[5;76H|[6;2H|[6;76H|[7;2H|[7;76H|[8;2H|[8;76H|[9;2H|[9;76H|[10;2H|[10;76H|[11;2H|[11;76H|[12;2H|[12;76H|[13;2H|[13;76H|[14;2H|[14;76H|[15;2H|[15;76H|[16;2H|[16;76H|[17;2H+-------------------------------------------------------------------------+[m

    Use the ^ and v keys to select which entry is highlighted.

    Press enter to boot the selected OS, 'e' to edit the

    commands before booting, or 'c' for a command-line.[5;78H [m[7m[5;3H Mirage                                                                  [5;75H[m[m[6;3H                                                                   [6;75H[m[m[7;3H                                                                   [7;75H[m[m[8;3H                                                                   [8;75H[m[m[9;3H                                                                   [9;75H[m[m[10;3H                                                                  [10;75H[m[m[11;3H                                                                  [11;75H[m[m[12;3H                                                                  [12;75H[m[m[13;3H                                                                  [13;75H[m[m[14;3H                                                                  [14;75H[m[m[15;3H                                                                  [15;75H[m[m[16;3H                                                                  [16;75H[m[16;78H [5;75H[23;4H The highlighted entry will be booted automatically in 1 seconds.   [5;75H[H[J  Booting 'Mirage'



root (hd0)

 Filesystem type is ext2fs, using whole disk

kernel /boot/mirage-os.gz



xc_dom_probe_bzimage_kernel: kernel is not a bzImage
close blk: backend at /local/domain/0/backend/vbd/2255/2049
close blk: backend at /local/domain/0/backend/vbd/2255/2064
Xen Minimal OS!
  start_info: 0000000000934000(VA)
    nr_pages: 0xf0000
  shared_inf: 0x7de83000(MA)
     pt_base: 0000000000937000(VA)
nr_pt_frames: 0x9
    mfn_list: 00000000001b4000(VA)
   mod_start: 0x0(VA)
     mod_len: 0
       flags: 0x0
    cmd_line: 
       stack: 00000000001935e0-00000000001b35e0
MM: Init
      _text: 0000000000000000(VA)
     _etext: 00000000000cb3df(VA)
   _erodata: 00000000000f4000(VA)
     _edata: 000000000015a340(VA)
stack start: 00000000001935e0(VA)
       _end: 00000000001b3a78(VA)
  start_pfn: 943
    max_pfn: f0000
Mapping memory range 0xc00000 - 0xf0000000
setting 0000000000000000-00000000000f4000 readonly
skipped 1000
MM: Initialise page allocator for 10c0000(10c0000)-f0000000(f0000000)
MM: done
Demand map pfns at f0001000-00000020f0001000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 00000000f0001000.
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(PATH) -> null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
getenv(OCAMLRUNPARAM) -> null
getenv(CAMLRUNPARAM) -> null
getenv(TMPDIR) -> null
getenv(TEMP) -> null
Netif: add resume hook
Netif.connect 1
Fatal error: exception Failure("net_1")
Raised at file "src/core/lwt.ml", line 788, characters 22-23
Called from file "lib/main.ml", line 58, characters 10-20
Called from file "main.ml", line 48, characters 2-28
Mirage exiting with status 2
Do_exit called!
base is 0x1aff20 caller is 0xb0b11
base is 0x159d00 caller is 0x0
base is 0xcd3e3 caller is 0x286e67696c616d65
base is 0x6d0065676172696d GPF rip: ca967, error_code=0
RIP: e030:[<00000000000ca967>] 
RSP: e02b:00000000001afee0  EFLAGS: 00010006
RAX: 000000000000001b RBX: 6d0065676172696d RCX: 0000000000001505
RDX: 0000000000001505 RSI: 00000000001afd30 RDI: 0000000000000004
RBP: 00000000001aff20 R08: 0000000000000504 R09: 0000000000000020
R10: 000000000000001b R11: 0000000000000000 R12: 00000020f0b9a040
R13: 0000000000001802 R14: 0000000000000000 R15: 0000000000000000
base is 0x1aff20 caller is 0xb0b11
base is 0x159d00 caller is 0x0
base is 0xcd3e3 caller is 0x286e67696c616d65
base is 0x6d0065676172696d GPF rip: cad57, error_code=0
RIP: e030:[<00000000000cad57>] 
RSP: e02b:00000000001afe08  EFLAGS: 00010016
RAX: 000000000000001b RBX: 6d0065676172696d RCX: 000000000000174e
RDX: 000000000000174e RSI: 00000000001afc58 RDI: 0000000000000004
RBP: 00000000001afe38 R08: 000000000000074d R09: 0000000000000020
R10: 000000000000001b R11: 0000000000000000 R12: 00000020f0b9a040
R13: 0000000000001802 R14: 0000000000000000 R15: 0000000000000000
base is 0x1afe38 caller is 0x0

1afdf0: 08 fe 1a 00 00 00 00 00 2b e0 00 00 00 00 00 00
1afe00: 57 ad 0c 00 00 00 00 00 03 00 00 00 30 00 00 00
1afe10: e0 fe 1a 00 00 00 00 00 6d 69 72 61 67 65 00 6d
1afe20: 20 ff 1a 00 00 00 00 00 40 a0 b9 f0 20 00 00 00

1afe20: 20 ff 1a 00 00 00 00 00 40 a0 b9 f0 20 00 00 00
1afe30: df af 0c 00 00 00 00 00 00 00 00 00 00 00 00 00
1afe40: 00 00 00 00 00 00 00 00 02 18 00 00 00 00 00 00
1afe50: 40 a0 b9 f0 20 00 00 00 20 ff 1a 00 00 00 00 00

cad40: 48 8b 5d 20 0f 1f 40 00 48 89 de 31 c0 bf a8 42
cad50: 0d 00 e8 d9 cf ff ff 48 8b 73 08 31 c0 bf b6 42
cad60: 0d 00 e8 c9 cf ff ff 48 8b 1b 48 85 db 75 d9 4c
cad70: 8b a5 98 00 00 00 49 81 fc ff 0f 00 00 76 61 49
```
It doesn't crash on alphanumeric ids. But Interfaces with alphanumeric ids always map to the first device (for a unikernel with multiple network devices--see implementation of connect), hence preventing TCP stack to work properly.

**Comment by mor1** [comment]
do you have the code that causes this crash?  (`main.ml` or `lib/main.ml` perhaps?)

**Comment by koleini** [comment]
```netif-forward``` in ```mirage-skeleton``` repo.

**Comment by koleini** [comment]
Can anyone compile and run ```netif-forward``` on a Xen machine? I am now having the same problem on Xen after recent updates.
@avsm @samoht @mor1

**Comment by mor1** [comment]
am building a xen virtualbox right now (as part of tidying up
`mirage-vagrant-vms`). will hopefully be able to try in that context this
afternoon...

On 19 March 2015 at 10:25, Masoud Koleini <notifications@github.com> wrote:

> Can anyone compile and run netif-forward on a Xen machine? I am now
> having the same problem on Xen after recent updates.
> @avsm <https://github.com/avsm> @samoht <https://github.com/samoht> @mor1
> <https://github.com/mor1>
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-net-xen/issues/22#issuecomment-83493138>
> .
>



-- 
Richard Mortier
mort@cantab.net

**Comment by yomimono** [comment]
Hi Masoud,

I'm able to run netif-forward (with a few changes, since `id` is no 
longer part of the DEVICE signature and ETHIF doesn't expose it 
directly) with alphanumeric interface names, although what the unikernel 
in fact receives is two netifs with the same MAC as you've reported.  If 
I change config.ml to ask for (netif "0") and (netif "1") instead of 
(netif "tap1") and (netif "tap2"), I get two working interfaces on the 
bridges specified.

This is on Xen x86 4.4, specifically the version packaged for Ubuntu 
14.04 (dpkg -s reports version 4.4.1-0ubuntu0.14.04.3 ).

Thanks,
Mindy

**Comment by koleini** [comment]
Thanks, I thought that my pull request that works with latest signatures is already merged.

@mor1: please use the code here:
https://github.com/mirage/mirage-skeleton/pull/78

**Comment by avsm** [comment]
@koleini -- that pull is waiting on a Travis pass, so you need to merge it back to the master branch and push the result to trigger it off.

**Comment by talex5** [comment]
Can this be closed? I think @yomimono's patch in 1.6.0 means the mistake will be detected in now.

**Comment by mor1** [comment]
Yes; closing
